File: 148.py
''' Given the head of a linked list, return the list after sorting it in ascending order.'''

class Solution(object):
    def sortList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if not head or not head.next:
            return head
        # Split the linked list into two halves
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        mid = slow.next
        slow.next = None
        left = self.sortList(head)
        right = self.sortList(mid)
        return self.merge(left, right)

    def merge(self, left, right):
        dummy = ListNode(0)
        tail = dummy
        while left and right:
            if left.val < right.val:
                tail.next = left
                left = left.next
            else:
                tail.next = right
                right = right.next
            tail = tail.next
        tail.next = left if left else right
        return dummy.next
    
        
        
         
        
        
        
        
        
        

File: 148.c
#include <stdlib.h>
#include <stdio.h>
#include <stdlib.h>

/*Given the head of a linked list, return the list after sorting it in ascending order*/
struct ListNode {
    int val;
    struct ListNode *next;
};
struct ListNode* sortList(struct ListNode* head){
    if (!head || !head->next) return head;

    // Find the middle point using the slow and fast pointer technique
    struct ListNode *slow = head, *fast = head, *prev = NULL;
    while (fast && fast->next) {
        prev = slow;
        slow = slow->next;
        fast = fast->next->next;
    }
    prev->next = NULL; // Split the list into two halves

    // Recursively sort the two halves
    struct ListNode *l1 = sortList(head);
    struct ListNode *l2 = sortList(slow);

    // Merge the two sorted halves
    struct ListNode dummy;
    dummy.next = NULL;
    struct ListNode *tail = &dummy;
    while (l1 && l2) {
        if (l1->val < l2->val) {
            tail->next = l1;
            l1 = l1->next;
        } else {
            tail->next = l2;
            l2 = l2->next;
        }
        tail = tail->next;
    }
    tail->next = l1 ? l1 : l2; // Append the remaining elements

    return dummy.next;
}